### PR TITLE
Implemented core

### PR DESCRIPTION
**Pull request by DarkLord017**

### CODE REVIEW

File: athena_abi/core.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+// @param abiname
+// @param classHash
+func StarknetAbiFromJSON(abiJson []map[string]interface{}, abiName string, classHash b256) (*StarknetABI, error) {
+	if abiJson == nil {
+		return nil, errors.New("unable to parse ABI")
+	}
+	groupedAbi := GroupAbiByType(abiJson)
+
+	// Parse defined types (structs and enums)
+	if groupedAbi["type_def"] == nil {
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: I don't think this is required.

File: athena_abi/core.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import (
+	"errors"
+	"log"
+)
+
+type b256 [32]byte
+
+type StarknetABI struct {
+	ABIName               *string
+	ClassHash             b256
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: I think we can simply go with []byte

File: athena_abi/core.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
+package athena_abi
+
+import (
+	"errors"
+	"log"
+)
+
+type b256 [32]byte
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: This might not be required. simple []byte should do the job.

File: athena_abi/core.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	if groupedAbi["type_def"] == nil {
+		log.Println("No type definitions found in ABI")
+	}
+	definedTypes, err := ParseEnumsAndStructs(groupedAbi["type_def"])
+	if err != nil {
+		sortedDefs, errDef := TopoSortTypeDefs(groupedAbi["type_def"])
+		if errDef != nil {
+			return nil, errors.New("unable to parse types")
+		}
+		defineTypes, errDtypes := ParseEnumsAndStructs(sortedDefs)
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: Can these not be combined into one statement ?

File: athena_abi/core.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		if errDtypes != nil {
+			return nil, errors.New("unable to parse types")
+		}
+		log.Println("ABI Struct and Enum definitions out of order & required topological sorting")
+	}
+
+	// Parse interfaces
+	var definedInterfaces []AbiInterface
+	if groupedAbi["interface"] == nil {
+		log.Println("No interfaces found in ABI")
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: Again not really required.

File: athena_abi/core.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		definedInterfaces = append(definedInterfaces, AbiInterface{
+			name:      iface["name"].(string),
+			functions: functions,
+		})
+	}
+
+	// Parse functions
+	if groupedAbi["function"] == nil {
+		log.Println("No functions found in ABI")
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: ditto

File: athena_abi/core.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+	for _, iface := range definedInterfaces {
+		for _, function := range iface.functions {
+			functions[function.name] = function
+		}
+	}
+
+	// Parse events
+	parsedAbiEvents := []AbiEvent{}
+	if groupedAbi["event"] == nil {
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: ditto.

File: athena_abi/core.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		log.Println("No events found in ABI")
+	}
+	for _, eventData := range groupedAbi["event"] {
+		parsedEvent, errParsingEvent := ParseAbiEvent(eventData, definedTypes)
+		if errParsingEvent != nil {
+			return nil, errors.New("unable to parse events")
+		}
+		parsedAbiEvents = append(parsedAbiEvents, *parsedEvent)
+	}
+	if len(parsedAbiEvents) == 0 {
```
Comment (Sh0g0-1758) [inline athena_abi/core.go]: ditto.

**Review by Sh0g0-1758 — CHANGES_REQUESTED** [review]
Code Quality in general is not so good. You should go through the code once more.

**Comment by DarkLord017** [comment]
Sir i tried to improve the code quality what can i improve more

**Review by Sh0g0-1758 — APPROVED** [review]
LGTM

**Pull request merged**